pkg/security/security.go

```diff
@@ -351,12 +352,42 @@ type Caller struct {
 	Identities []string
 }
 
+// AuthContext carries the contextual information and is passed across authenticators.
+type AuthContext struct {
+	// RequestContext is the context from request.
+	RequestContext context.Context
+	// Authenticators is the list of authenticators that were executed before in the auth chain.
+	Authenticators []string
+	// Caller indicates the caller that made authentication request.
+	Caller *Caller
+}
+
+// Authenticator determines the caller identity based on request context.
 type Authenticator interface {
-	Authenticate(ctx context.Context) (*Caller, error)
+	Authenticate(ctx *AuthContext) (*Caller, error)
 	AuthenticatorType() string
 	AuthenticateRequest(req *http.Request) (*Caller, error)
 }
 
+// DelegatedAuthenticator determines whether the main authenticator can trust the
+// client and proceed with authentication. This is used in cases where the client
+// authentication can not be directly determined from the request(no TLS certs, JWT etc)
+// but request has additional information to authenticate.
+// For example XfccAuthenticators should validate whether client is from trusted Cidr range
+// before it can trust the Xfcc Header.
+type DelegatedAuthenticator interface {
```

Comment:
I would not define an interface for this. Just code the logic in the authenticator. There are all kind of logical issues here - you may trust the caller but not as an authenticating/delegating proxy, etc.

Too early to abstract this - we have a single case so far.